tools/wake-format/emitter.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include "emitter.h"
+
+#include <cassert>
+
+#include "parser/parser.h"
+
+wcl::rope Emitter::newline(ctx_t ctx) {
+  wcl::rope_builder builder;
+
+  builder.append("⏎\n");
+  for (int i = 0; i < ctx.nest_level; i++) {
+    builder.append(space(space_per_indent));
+  }
+
+  return std::move(builder).build();
+}
+
+wcl::rope Emitter::space(uint8_t count) {
+  wcl::rope_builder builder;
+  for (uint8_t i = 0; i < count; i++) {
+    builder.append("·");
+  }
+  return std::move(builder).build();
+}
+
+wcl::optional<wcl::rope> Emitter::flat(ctx_t ctx, CSTElement node) {
+  return walk_node(ctx.flat(), node);
+}
+
+wcl::rope Emitter::try_flat(ctx_t ctx, CSTElement node) {
+  auto flat_node = flat(ctx, node);
+  if (flat_node && flat_node->size() + ctx.nest_level * space_per_indent < 80) {
```

Comment:
Shouldn't this be worried about the current column? I think your ctx needs to know the current column in order to figure this out.

Reply:
Right, this isn't going to correctly determine column width. this is mostly just to test that _really_ long lines are forced to be split. This will need to be fixed but I think it can be deferred for the moment

Reply:
hmm, ok lets chat about your more complete local code today. I think this is a fundamental issue we need to make sure we solve.

Reply:
Please make the column changes we discussed to make rope track column and fix this line

nit: Also let's make the line length a variable and set it to 100 not 80. I did a survey of wake code a while back and found that 120 is probably better for wake code though since we hit that today in practice even with heavy indentation since somethings like the first line of a def can't be broken up right now.